src/assume.h

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_ASSUME_H
+#define BITCOIN_ASSUME_H
+
+#include <tinyformat.h>
+
+// ASSUME(expression)
+// ==================
+// ASSUME(...) is used to document assumptions.
+//
+// Bitcoin Core is always compiled with assertions enabled. assert(...)
+// is thus only appropriate for documenting critical assumptions
+// where a failed assumption should lead to immediate abortion also
+// in production environments.
+//
+// ASSUME(...) works like assert(...) if -DABORT_ON_FAILED_ASSUME
+// and is side-effect safe.
+//
+// More specifically:
+//
+// * If compiled with -DABORT_ON_FAILED_ASSUME (set by --enable-debug):
+//   Evaluate expression and abort if expression is falsy.
+//
+// * Otherwise:
+//   Evaluate expression and continue execution.
+//
+// ABORT_ON_FAILED_ASSUME should not be set in production environments.
+//
+// Example
+// =======
+// int main(void) {
+//     ASSUME(IsFoo());
+//     ...
+// }
+//
+// If !IsFoo() and -DABORT_ON_FAILED_ASSUME, then:
+//     filename.cpp:123: main: ASSUME(IsFoo()) failed.
+//     Aborted
+// Otherwise the execution continues.
+
+#ifdef ABORT_ON_FAILED_ASSUME
+#define ASSUME_FAILED(expression, file, line, func)                                                  \
+    do {                                                                                             \
+        tfm::format(std::cerr, "%s:%d: %s: ASSUME(%s) failed. You may report this issue here: %s\n", \
+            file, line, func, expression, PACKAGE_BUGREPORT);                                        \
+        std::abort();                                                                                \
+    } while (false)
+#else
+#define ASSUME_FAILED(expression, file, line, func) ((void)0)
+#endif
```

Comment:
I'm still not sure it wouldn't be better unconditionally log (independent of compile flags) something like `%s:%d: ASSUME(%s) failed` and change the testing harnesses to emit a failure if that happens? If we continue fine when running release versions anyway, wouldn't it make more sense to capture all the failures while testing to reduce the number of restarts needed?

My inclination would be more along the lines of:

```
if (!(expression)) {
    LogPrintf("Assumption failed %s:%d: %s", __FILE__,__LINE__, #expression);
    if (!Params().IsTestChain()) std::abort();
}
```

(or have the `std::abort()` always happen if a compile time flag is set, if that's helpful for static analysis)

Reply:
has been fixed, I believe